cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,421 @@
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if self.distance(p, q) > self._control_radius:
+                            raise ValueError("Qubits {!r}, {!r} are too "
+                                             "far away".format(p, q))
+            return
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(operation.gate, (ops.XPowGate,
+                                       ops.YPowGate,
+                                       ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubits)):
+                raise ValueError("Bad number of XY gates in parallel")
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        GATE_CATEGORIES = {
+            ops.ZPowGate: 'Z',
+            ops.pauli_gates._PauliZ: 'Z',
+
+            ops.pauli_gates._PauliX: 'XY',
+            ops.pauli_gates._PauliY: 'XY',
+            ops.XPowGate: 'XY',
+            ops.YPowGate: 'XY',
+            ops.PhasedXPowGate: 'XY',
+
+            ops.CNotPowGate: 'controlled',
+            ops.CZPowGate: 'controlled',
+            ops.CCXPowGate: 'controlled',
+            ops.CCZPowGate: 'controlled',
+
+            ops.MeasurementGate: 'measure',
+        }
+
+        operations = moment.operations
+
+        categorized_ops = collections.defaultdict(list) #type: DefaultDict
+        for op in operations:
+            assert isinstance(op,
+                              (ops.GateOperation, ops.ParallelGateOperation))
+            categorized_ops[GATE_CATEGORIES[type(op.gate)]].append(op)
+
+        for k in ['Z', 'XY', 'controlled']:
+            if len(set(op.gate for op in categorized_ops[k])) > 1:
+                raise ValueError(
+                    "Non-identical simultaneous {} gates".format(k))
+
+        num_parallel_xy = sum([len(op.qubits) for op in categorized_ops['XY']])
+        num_parallel_z = sum([len(op.qubits) for op in categorized_ops['Z']])
+        has_measurement = len(categorized_ops['measure']) > 0
+        controlled_qubits_lists = [op.qubits
+                                   for op in categorized_ops['controlled']]
+        # htype: List[Tuple[raw_types.Qid, ...]]
+        if len(controlled_qubits_lists) > 0:
```

Comment:
Merge this condition into the three sub-conditions, instead of having the block.

The first and third sub-conditions imply this, so they don't need to be changed. The second condition can be changed to `controlled_qubits_lists and (num_parallel_xy or num_parallel_z)`. This rewrite uses the fact that the empty list and 0 are falsy (and in fact the style guide says to prefer `if x:` to `if len(x) > 0`).

Reply:
done